### DISCUSSION TITLE
Supabase won't redirect to MAUI Blazor App

### DISCUSSION THREAD

**Discussion by Dayjay** [dotMorten/WinUIEx#140] [Q&A]
Hey,

I am using Supabase and WinUIEx for the authentication in my MAUI Blazor Hybrid App. This is the code I am using:

```csharp
var state = await Supabase.Auth.SignIn(Constants.Provider.Azure, new SignInOptions
    {
        FlowType = Constants.OAuthFlowType.PKCE,
        RedirectTo = "myclient://oauth/callback"
    });

var result = await WinUIEx.WebAuthenticator.AuthenticateAsync(state.Uri, new Uri("myclient://oauth/callback"));

//do stuff with the result - but I never get here
```
When I call the code it opens the browser window (tried with Firefox and Edge) but I end up with the following error:

![grafik](https://github.com/dotMorten/WinUIEx/assets/25659168/06f89641-bb44-4e56-8034-ddf8341edd94)

I tried pasting the `state.Uri` in the browser and it works, it even opens my maui app (again since its already open).
I also added the url as a valid redirect url in the Supabase dashboard.

Is there something I am missing?

**Comment by dotMorten** [comment]
> I end up with the following error:

Where is this error happening? What URL is the browser trying to open? How is that different from state.Uri ?

Wrt Firefox, see https://github.com/dotMorten/WinUIEx/issues/117

**Comment by Dayjay** [comment]
Hey, 

This is the URL (=state.Uri) it opens in the browser and that I just pasted:

`https://1234567890.supabase.co/auth/v1/authorize?flow_type=pkce&code_challenge=dskhflsdkgaljfflashfkjsgladas&code_challenge_method=s256&provider=azure&redirect_to=myclient%3a%2f%2foauth%2fcallback`

The error I posted is the response I get in the browser from the supabase API. Doesn't matter which browser.
The error only appears if I use the WinUIEx.WebAuthenticator. 

Even if I use `NavigationManager.NavigateTo()` it opens the Browser window fine and opens the app again.

Just to clarify the workflow:

- Login Button gets clicked
- Calls the Supabase URL I posted above
- Supabase redirects to my provider (in my case Azure)
- Azure redirects back to Supabase
- Supabase should redirect back to my app (this is when I get the error)

Honestly wouldn't be surprised if this is just another annoying MAUI Bug somehow.

**Reply by dotMorten** [reply]
That url seems to be missing the state parameter, and my guess is supabase doesn’t know how to handle a state parameter correctly

**Reply by Dayjay** [reply]
Certainly looks like it, I found a similiar issue when someone tried using Github with Supabase.

Thanks for the quick response, I will figure something out.

**Reply by dotMorten** [reply]
Any idea why the state parameter isn’t in the above url? The url you pasted doesn’t look like something WinUIEx would create

**Reply by Dayjay** [reply]
Did a bit more digging now, WinUIEx does actually add the state parameter. But Supabase doesnt seem to like it hence why the URL above works when just pasting it in the browser.

![grafik](https://github.com/dotMorten/WinUIEx/assets/25659168/52fce0d7-d7a1-4533-b19b-60b8c81d0ede)

It even still exists when I get the code back from supabase. 

![grafik](https://github.com/dotMorten/WinUIEx/assets/25659168/9dd38abe-b5a6-49c2-befc-30ac88946885)

But for some reason I just get a Bad Request (400) Error with the message I posted above.

**Reply by dotMorten** [reply]
First of all the app instance is parameter is empty. That shouldn’t happen. Secondly it’s not url encoded. The latter makes me think it’s supabase - I’ve seen other services doing this wrong but try latest 2.3.2 since that one  encodes the url a bit better

**Reply by Dayjay** [reply]
The empty appInstanceId I could fix with 
```csharp
Microsoft.Windows.AppLifecycle.AppInstance.FindOrRegisterForKey("myAppId");
```
That seems to be a MAUI issue.
I tested with 2.3.2 and the state is not url encoded.

**Reply by dotMorten** [reply]
WinUIEx should be doing that for you. Did you add the oauth redirection check to the app launch as documented?

**Reply by Dayjay** [reply]
Yes, I did. This is how my `Windows/App.xaml.cs` looks like:

```csharp
public App()
{
    if (WinUIEx.WebAuthenticator.CheckOAuthRedirectionActivation())
        return;
    this.InitializeComponent();
}

protected override MauiApp CreateMauiApp() => MauiProgram.CreateMauiApp();
```

Edit: uri-scheme is also registered.

**Reply by dotMorten** [reply]
Sorry I don't have much more to go on. It might be worth cloning this repo, and change the oauth sample page to use your service, and just step through winuiex and see if you can spot what's going wrong, compared to how for instance Maui's WebAuthenticator is behaving.

**Reply by Dayjay** [reply]
Will do and report back later. 

But thank you very much. 👍

**Reply by Dayjay** [reply]
Thinking about it, I bet it breaks because the WebAuthenticator runs inside the BlazorWebView MAUI Hybrid creates and not the actual Maui Thread.

Would also explain why the instance id is emtpy and why other stuff maybe fails.

**Reply by dotMorten** [reply]
No idea. I’ve never used Maui blazor. But try the sample app to confirm if it works outside blazor

**Reply by Dayjay** [reply]
Tried different things over the weekend with Hybrid MAUI but no success.

In the `Platforms/Windows/App.xaml`, if I understood it correct, it should at least set a guid as key but not even that is happening:

![grafik](https://github.com/dotMorten/WinUIEx/assets/25659168/eb0b9abc-34b7-400d-afce-65ff6742d400)

`activatedEventArgs` is not null. However, the Kind is `ExtendedActivationKind.Launch` and not `ExtendedActivationKind.Protocol`.

Edit: Forget what I said, I think I misunderstood what the function does.

**Comment by Dayjay** [comment]
I got it working. :)

I copied your WebAuthenticator to my project and adjusted it so it would work with the Supabase API.

Mainly:

I don't pass the state parameter anymore - supabase just doesn't like it. It's a bit problematic because I don't know the Id of the task anymore in the callback but for my use case it shouldn't be a problem (I think).

The appInstanceId can always be the same and in the `ResumeSignin` I just get the first task and remove it before I set the result.
